src/compose/volume.ts

```diff
@@ -0,0 +1,143 @@
+import { InternalInconsistencyError } from '../lib/errors';
+import * as LogTypes from '../lib/log-types';
+import { LabelObject } from '../lib/types';
+import Logger from '../logger';
+import * as ComposeUtils from './utils';
+
+export interface VolumeConstructOpts {
+	logger: Logger;
+	docker: Docker;
+}
+
+export interface VolumeConfig {
+	labels: LabelObject;
+	driverOpts: Docker.VolumeInspectInfo['Options'];
+}
+
+export class Volume {
+	public appId: number;
+	public name: string;
+	public config: VolumeConfig;
+
+	private logger: Logger;
+	private docker: Docker;
+
+	private constructor(
+		name: string,
+		appId: number,
+		config: VolumeConfig,
+		opts: VolumeConstructOpts,
+	) {
+		this.name = name;
+		this.appId = appId;
+
+		this.logger = opts.logger;
+		this.docker = opts.docker;
+		this.config = config;
+	}
+
+	public static fromDockerVolume(
+		opts: VolumeConstructOpts,
+		inspect: Docker.VolumeInspectInfo,
+	): Volume {
+		// Convert the docker inspect to the config
+		const config: VolumeConfig = {
+			labels: inspect.Labels || {},
+			driverOpts: inspect.Options || {},
+		};
+
+		// Detect the name and appId from the inspect data
+		const { name, appId } = this.deconstructDockerName(inspect.Name);
+
+		return new Volume(name, appId, config, opts);
+	}
+
+	public static fromComposeObject(
+		name: string,
+		appId: number,
+		config: Partial<VolumeConfig>,
+		opts: VolumeConstructOpts,
+	) {
+		const filledConfig: VolumeConfig = {
+			driverOpts: config.driverOpts || {},
+			labels: ComposeUtils.normalizeLabels(config.labels || {}),
+		};
+
+		// We only need to assign the labels here, as when we
+		// get it from the daemon, they should already be there
+		assign(filledConfig.labels, constants.defaultVolumeLabels);
+
+		return new Volume(name, appId, filledConfig, opts);
+	}
+
+	public isEqualConfig(volume: Volume): boolean {
+		return (
+			isEqual(this.config.driverOpts, volume.config.driverOpts) &&
+			isEqual(this.config.labels, volume.config.labels)
+		);
+	}
+
+	public async create(): Promise<void> {
+		this.logger.logSystemEvent(LogTypes.createVolume, {
+			volume: { name: this.name },
+		});
+		// Check that we're not trying to recreate a volume that
+		// already exists
+		await this.docker.createVolume({
+			Name: Volume.generateDockerName(this.appId, this.name),
+			Labels: this.config.labels,
+			DriverOpts: this.config.driverOpts,
+		});
+	}
+
+	public async remove(): Promise<void> {
+		this.logger.logSystemEvent(LogTypes.removeVolume, {
+			volume: { name: this.name },
+		});
+
+		try {
+			await this.docker
+				.getVolume(Volume.generateDockerName(this.appId, this.name))
+				.remove();
+		} catch (e) {
+			this.logger.logSystemEvent(LogTypes.removeVolumeError, {
+				volume: { name: this.name, appId: this.appId },
+				error: e,
+			});
+		}
+	}
+
+	public static generateDockerName(appId: number, name: string) {
+		return `${appId}_${name}`;
+	}
+
+	private static deconstructDockerName(
```

Comment:
Getting the network would need to be changes to a "list networks with label filter" call in such case.

Reply:
Me too, but the same problem applies as here: https://github.com/balena-io/balena-supervisor/pull/1028#discussion_r299444489, I do plan to move to this kind of architecture, but I want to make sure that we're not going to delete any user data.

Reply:
Note that having the names defined like this is also useful to easily reference them when creating service containers - otherwise we'd also have to find the correct name for an app's volume when including it in each container. And I made it so that we also get these properties (appId, volume name) from the name so that there's no duplication.

Reply:
It's 100% useful to have such names, but I don't think supervisor logic should rely on name format: we might want to change it. 
To me, it makes sense to have friendly names helping with support/triaging/debugging, but not rely on decoding metadata from them. It would be easier on the long term to rely only on labels for structured metadata.